src/routes/product/[id([0-9]+)]/edit.svelte

```diff
@@ -0,0 +1,352 @@
+      if (product.sized) {
+        currentVariety.sizes[varieties[i].size] = varieties[i].amount;
+      }
+    }
+    product.varieties.push(currentVariety);
+  }
+
+  export let colors;
+  export let sizes;
+  export let account;
+  $session.user = account;
+
+  let errors = {
+    name: false,
+    price: false,
+  };
+  let errorMessage = null;
+  let warningDialogOpen = false;
+
+  async function addColor(color) {
+    let resp = await api.post('/colors', { data: { value: color } });
+    if (resp.status == 200) {
+      colors.push({ value: color });
+      colors = colors;
+    } else {
+      console.error(`Color creation failed (${resp.status})`);
+    }
+  }
+
+  function changeProductField({ field, value }) {
+    product[field] = value;
+    if (errors[field]) {
+      errors[field] = false;
+    }
+  }
+
+  function deleteProduct() {
+    prefetch('/store');
+    warningDialogOpen = false;
+    api.del(`/products/${unmodifiedProduct.id}`).then(resp => {
+      if (resp.ok) {
+        return goto('/store');
+      }
+
+      if (resp.status === 400) {
+        resp.json().then(message => {
+          if ('message' in message) {
+            errorMessage = JSON.stringify(message.message);
+          } else {
+            errorMessage = JSON.stringify(message);
+          }
+          console.error(message);
+        });
+      } else {
+        errorMessage = 'The universe really wants this product. Try deleting again later.';
+        resp.text().then(console.error);
+      }
+    });
+  }
+
+  const unwrapValue = (obj) => obj.value;
+
+  function closeWarningDialog() {
+    warningDialogOpen = false;
+  }
+
+  function createVariety() {
+    product.varieties.push(getBlankVariety());
+    product = product;
+  }
+
+  function removeVariety(e) {
+    product.varieties.splice(e.detail, 1);
+    product = product;
+  }
+
+  function changeVariety(e) {
+    const { index, variety } = e.detail;
+    product.varieties[index] = variety;
+  }
+
+  function addNewFile({ detail }) {
+    product.varieties[detail.varietyIndex].images[detail.fileIndex] = `/file/${detail.id}`;
+  }
+
+  function deleteFile({ detail }) {
+    product.varieties[detail.varietyIndex].images.splice(detail.fileIndex, 1);
+    product = product;
+  }
+
+  function swapFiles({ detail }) {
+    let images = product.varieties[detail.varietyIndex].images;
+    let temp = images[detail.oldDraggableIndex];
+    images[detail.oldDraggableIndex] = images[detail.newDraggableIndex];
+    images[detail.newDraggableIndex] = temp;
+    product = product;
+  }
+
+  function computeDiffs(updated, old) {
+    const diffs = {
+      fields: null,
+      varieties: [],
+      deletedVarieties: null,
+    };
+
+    for (let field of ['name', 'type', 'description', 'price']) {
+      if (updated[field] !== old[field]) {
+        if (diffs.fields == null) {
+          diffs.fields = { [field]: updated[field] };
+        } else {
+          diffs.fields[field] = updated[field];
+        }
```

Comment:
Doesn't look so idiomatic. Suggestion:
```js
if (diffs.fields == null)
    diffs.fields = {};
diffs.fields[field] = updated[field];
```

Reply:
Thought about that, but idk which way really is more idiomatic for JS. The literal way seemed like what I'd opt for in, say, Python, so I went with it here. With your permission, I'll keep it as it is

Reply:
You may keep it if you don't want to dismiss the approvals, but that style is not so readable